File: demo/c/can_frame_demo.c
/*
 * can_frame_demo.c — OneroArm 纯 C ABI：原始 CAN 帧收发最小示例
 *
 * 演示原始 CAN 帧四个接口：
 *   - onero_register_can_frame_callback：注册接收回调（仅非保留 ID 派发）
 *   - onero_send_can_frame              ：向**非保留** ID 发送一帧
 *   - onero_pump_can_bus                ：主动驱动一次串口 rx，把帧从 SLCAN 缓冲取出
 *   - onero_clear_can_frame_callback    ：清回调
 *
 * 注意：
 *   - 本 demo **不调用 enable_motors** —— 串口在 onero_create_robot 时已打开，
 *     收发原始帧不需要电机使能。
 *   - 选用 CAN ID = 0x100，不在 SDK 保留集 (0x01-0x08, 0x11-0x17, 0x7FF) 内，
 *     不会被 ONERO_ERR_RAW_FRAME_RESERVED_ID 拦截。
 *   - 总线上若**没有**接听该 ID 的节点，回调可能不会触发；这是预期行为。
 */
#include "onero_interface_c.h"
#include <stdio.h>
#include <string.h>

/* === 测试帧 ============================== */
static const uint16_t TEST_CAN_ID  = 0x100;                       /* 非保留 */
static const uint8_t  TEST_PAYLOAD[] = {0xDE, 0xAD, 0xBE, 0xEF};
static const int      PUMP_TIMES     = 5;
static const int      PUMP_TIMEOUT_MS = 50;
/* ========================================= */

static onero_config_t build_config(void) {
    onero_config_t cfg;
    memset(&cfg, 0, sizeof(cfg));
    strncpy(cfg.device,      "/dev/ttyACM0", sizeof(cfg.device)      - 1);
    strncpy(cfg.robot_model, "a1_l",         sizeof(cfg.robot_model) - 1);
    strncpy(cfg.version,     "A1",           sizeof(cfg.version)     - 1);
    strncpy(cfg.mount_orientation, "vertical", sizeof(cfg.mount_orientation) - 1);
    cfg.dof = 7;
    cfg.baud_rate = 921600;
    return cfg;
}

static void on_can_frame(uint16_t can_id,
                         const uint8_t* data, uint8_t len,
                         void* user_data) {
    (void)user_data;
    printf("[RX] id=0x%03X len=%u data=", can_id, (unsigned)len);
    for (uint8_t i = 0; i < len; ++i) printf("%02X ", data[i]);
    printf("\n");
}

int main(void) {
    onero_config_t cfg = build_config();
    onero_handle h = onero_create_robot(&cfg);
    if (!h) {
        fprintf(stderr, "create_robot failed\n");
        return 1;
    }

    int rc = onero_register_can_frame_callback(h, on_can_frame, NULL);
    if (rc != 0) {
        fprintf(stderr, "register_can_frame_callback failed, ret=%d\n", rc);
    } else {
        printf("[OK] register_can_frame_callback\n");
    }

    printf("[TX] id=0x%03X len=%zu data=", TEST_CAN_ID, sizeof(TEST_PAYLOAD));
    for (size_t i = 0; i < sizeof(TEST_PAYLOAD); ++i) printf("%02X ", TEST_PAYLOAD[i]);
    printf("\n");
    rc = onero_send_can_frame(h, TEST_CAN_ID, TEST_PAYLOAD, sizeof(TEST_PAYLOAD));
    if (rc != 0) {
        fprintf(stderr, "[X] send_can_frame failed, ret=%d (无硬件预期 -13)\n", rc);
    } else {
        printf("[OK] send_can_frame\n");
    }

    /* 主动驱动 rx，让总线上的回帧（如有）经回调派发 */
    for (int i = 0; i < PUMP_TIMES; ++i) {
        onero_pump_can_bus(h, PUMP_TIMEOUT_MS);
    }

    onero_clear_can_frame_callback(h);
    onero_destroy_robot(h);
    return 0;
}

File: demo/python/can_frame_demo.py
"""can_frame_demo.py — OneroArm Python：原始 CAN 帧收发最小示例

演示原始 CAN 帧四个接口：
  - register_can_frame_callback：注册接收回调（仅非保留 ID 派发）
  - send_can_frame              ：向**非保留** ID 发送一帧
  - pump_can_bus                ：主动驱动一次串口 rx，把帧从 SLCAN 缓冲取出
  - clear_can_frame_callback    ：清回调

注意：
  - 本 demo **不调用 enable_motors** —— 串口在 OneroArm 构造时已打开，
    收发原始帧不需要电机使能。
  - 选用 CAN ID = 0x100，不在 SDK 保留集 (0x01-0x08, 0x11-0x17, 0x7FF) 内，
    不会被 ONERO_ERR_RAW_FRAME_RESERVED_ID 拦截。
  - 总线上若**没有**接听该 ID 的节点，回调可能不会触发；这是预期行为。
"""
import threading
import time

import oneroarm

# === 测试帧 ===============================
TEST_CAN_ID  = 0x100              # 非保留
TEST_PAYLOAD = bytes([0xDE, 0xAD, 0xBE, 0xEF])
PUMP_STEP_MS         = 50         # 单次 pump 步长；越小越能及时跳出阻塞
RECV_OVERALL_TIMEOUT = 5.0        # 秒：等待回帧的硬超时上限（无节点时必触发）
# =========================================

# 回调内禁止重入 arm 的 send_* / 运动控制方法，所以仅置位一个 Event，
# 让外层 pump 循环据此判断是否退出阻塞。
_recv_event = threading.Event()


def build_config() -> oneroarm.OneroConfig:
    cfg = oneroarm.OneroConfig()
    cfg.device = "/dev/ttyACM0"
    cfg.robot_model = "a1_l"
    cfg.version = "A1"
    cfg.mount_orientation = "vertical"
    return cfg


def on_can_frame(can_id: int, data: bytes) -> None:
    print(f"[RX] id=0x{can_id:03X} len={len(data)} data={data.hex(' ').upper()}")
    _recv_event.set()


def main() -> None:
    cfg = build_config()
    arm = oneroarm.OneroArm(cfg)

    rc = arm.register_can_frame_callback(on_can_frame)
    if rc != 0:
        print(f"register_can_frame_callback failed, ret={rc}")
    else:
        print("[OK] register_can_frame_callback")

    print(f"[TX] id=0x{TEST_CAN_ID:03X} len={len(TEST_PAYLOAD)} "
          f"data={TEST_PAYLOAD.hex(' ').upper()}")
    rc = arm.send_can_frame(TEST_CAN_ID, TEST_PAYLOAD)
    if rc != 0:
        print(f"[X] send_can_frame failed, ret={rc} (无硬件预期 -13)")
    else:
        print("[OK] send_can_frame")

    # 阻塞至收到任意非保留 ID 的 CAN 帧，或 RECV_OVERALL_TIMEOUT 到期
    deadline = time.monotonic() + RECV_OVERALL_TIMEOUT
    while not _recv_event.is_set() and time.monotonic() < deadline:
        arm.pump_can_bus(PUMP_STEP_MS)

    if _recv_event.is_set():
        print("[OK] 收到回帧")
    else:
        print(f"[X] {RECV_OVERALL_TIMEOUT:.1f}s 内未收到任何 CAN 帧")

    arm.clear_can_frame_callback()


if __name__ == "__main__":
    main()
